lib/Parse/ParseExpr.cpp

```diff
@@ -2399,6 +2399,41 @@ parseClosureSignatureIfPresent(SmallVectorImpl<CaptureListEntry> &captureList,
     }
   }
 
+  if (!params)
+    return invalid;
+
+  // If this was a closure declaration (maybe even trailing)
+  // tuple parameter destructuring is one of the common
+  // problems, and is misleading to users, so it's imperative
+  // to detect any tuple splat or destructuring as early as
+  // possible and give a proper fix-it. See SE-0110 for more details.
+  for (unsigned i = 0, e = params->size(); i != e; ++i) {
+    auto *param = params->get(i);
+    if (!param->isInvalid())
+      continue;
+
+    auto paramRange = CharSourceRange(
+        SourceMgr, param->getStartLoc(),
+        Lexer::getLocForEndOfToken(SourceMgr, param->getEndLoc()));
+
+    auto argName = "arg" + std::to_string(i);
+
+    SmallString<64> fixIt;
+    llvm::raw_svector_ostream OS(fixIt);
+
+    if (Tok.isAtStartOfLine())
+      OS << '\n';
+
+    OS << "let " << SourceMgr.extractText(paramRange) << " = " << argName
+       << (Tok.isAtStartOfLine() ? "\n" : "; ");
```

Comment:
It would be nice to also properly indent the *next* line as well. It looks like Lexer::getIndentationForLine can do that.